src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,253 @@
+	waitJitter	= 0.5
+	waitSteps	= 5
+	waitCap		= time.Minute * 5
+)
+
+var backoff = wait.Backoff{
+	Duration:	waitDuration,
+	Factor:		waitFactor,
+	Jitter:		waitJitter,
+	Steps:		waitSteps,
+	Cap:		waitCap,
+}
+
+func (r *StatefulSetResource) updateStsImage(
+	ctx context.Context, sts *appsv1.StatefulSet,
+) error {
+	rpContainer, err := findContainer(sts.Spec.Template.Spec.Containers, redpandaContainerName)
+	if err != nil {
+		return err
+	}
+
+	newImage := fmt.Sprintf("%s:%s", r.pandaCluster.Spec.Image, r.pandaCluster.Spec.Version)
+	if rpContainer.Image != newImage {
+		r.logger.Info("Image has changed", "cluster image", newImage, "container image", rpContainer.Image)
+
+		stsSpec := sts.Spec.Template.Spec
+
+		if err := modifyContainerImage(stsSpec.InitContainers, configuratorContainerName, newImage); err != nil {
+			return err
+		}
+
+		if err := modifyContainerImage(stsSpec.Containers, redpandaContainerName, newImage); err != nil {
+			return err
+		}
+
+		if err := r.partitionUpdateImage(ctx, sts, newImage); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	r.logger.Info("Start statefulset image update")
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := *sts.Spec.Replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
+			continue
+		}
+
+		// Continue only if error is due to pod not ready, or unchanged image.
+		if !errors.Is(poderr, errContainerHasWrongImage) && !errors.Is(poderr, errPodNotReady) {
+			return poderr
+		}
+
+		// Ideally, at this point we'd wait until all Pods are in ready state.
+		// However, if something went wrong with a previous image change and the pod
+		// is in bad state, e.g., due to wrong or erroneous image, the pod will
+		// keep restarting and require manual intervention. Instead, for now we
+		// skip the readiness requirement and "force" the image update.
+
+		if err := r.rollingUpdatePartition(ctx, sts, ordinal); err != nil {
+			return err
+		}
+
+		r.logger.Info("Waiting for pod to complete updating", "partition:", ordinal)
+
+		if err := r.waitUntilPodImageIsVerified(ctx, sts, newImage, ordinal); err != nil {
+			return err
+		}
+
+		// Give time to redpanda pod to rejoin cluster.
+		// In a revision, an API call to redpanda is made to help us decide when to continue.
+		r.logger.Info("Sleeping for 1 minute before updating the next pod, or completion")
+		time.Sleep(time.Minute)
```

Comment:
oh wow, I would really like to make this asynchronous somehow...

Reply:
we should absolutely at least have some timeout here but then when we hit that timeout, we let the sts in kind of weird state so we need to make sure that the upgrade phase can be continued after that...

Reply:
Yeah, this will be removed (I thought I did already), and be replaced with logic towards the start of the upgrade. I'll update the commit these days.

Reply:
@alenkacz I updated the commit. There's no sleep() anymore. I also added a call to the broker's Kafka API that I will update in another commit to check the previously updated pod has joined all its groups. This goes beyond the "READY" state that k8s provides. (I don't add it as a condition to the pod's status because we don't always want it to be a condition for a pod to be ready).